stan/math/prim/mat/prob/normal_id_glm_lpdf.hpp

```diff
@@ -0,0 +1,162 @@
+     * @tparam T_x type of the matrix of independent variables (features); this
+     * should be an Eigen::Matrix type whose number of rows should match the
+     * length of n and whose number of columns should match the length of beta
+     * @tparam T_beta type of the weight vector;
+     * this can also be a single value;
+     * @tparam T_alpha type of the intercept;
+     * this has to be a single value;
+     * @tparam T_scale type of the scale vector.
+     * @param n vector parameter
+     * @param x design matrix
+     * @param beta weight vector
+     * @param alpha intercept (in log odds)
+     * @param sigma (Sequence of) scale parameters for the normal
+     * distribution.
+     * @return log probability or log sum of probabilities
+     * @throw std::domain_error if x, beta or alpha is infinite.
+     * @throw std::domain_error if the scale is not positive.
+     * @throw std::invalid_argument if container sizes mismatch.
+     */
+    template <bool propto, typename T_n, typename T_x, typename T_beta,
+              typename T_alpha, typename T_scale>
+    typename return_type<T_n, T_x, T_beta, T_alpha, T_scale>::type
+    normal_id_glm_lpdf(const T_n &n, const T_x &x, const T_beta &beta,
+                             const T_alpha &alpha, const T_scale& sigma) {
+      static const std::string function = "normal_id_glm_lpdf";
+      typedef typename stan::partials_return_type<T_n, T_x, T_beta,
+                                                  T_alpha, T_scale>::type
+        T_partials_return;
+
+      using std::exp;
+      using Eigen::Dynamic;
+      using Eigen::Matrix;
+      using Eigen::Array;
+
+      if (!(stan::length(n) && stan::length(x) && stan::length(beta)
+        && stan::length(sigma)))
+        return 0.0;
+
+      T_partials_return logp(0.0);
+
+      check_finite(function, "Vector of dependent variables", n);
+      check_finite(function, "Matrix of independent variables", x);
+      check_finite(function, "Weight vector", beta);
+      check_finite(function, "Intercept", alpha);
+      check_positive(function, "Scale vector", sigma);
+      check_consistent_sizes(function,
+                             "Rows in matrix of independent variables",
+                             x.col(0), "Vector of dependent variables",  n);
+      check_consistent_sizes(function,
+                             "Rows in matrix of independent variables",
+                             x.col(0), "Vector of scale paramters",  sigma);
+      check_consistent_sizes(function,
+                             "Columns in matrix of independent variables",
+                             x.row(0), "Weight vector",  beta);
+
+      if (!include_summand<propto, T_n, T_x, T_beta, T_alpha, T_scale>::value)
+        return 0.0;
+
+      const size_t N = x.col(0).size();
+      const size_t M = x.row(0).size();
+
+      Array<T_partials_return, Dynamic, 1> sigma_dbl(N, 1);
+      Array<T_partials_return, Dynamic, 1> n_dbl(N, 1);
```

Comment:
I'm sure this is right but for my own edification, why do you use `Eigen::Array` here vs. Vector?

Reply:
As far as I understand it Eigen::Arrays and Eigen::Vectors/Eigen::Matrices are internally represented the same way, but they provide a different interface. The former provide an interface for coefficientwise operations (like coefficientwise exponential) while the latter provide an interface for matrix/vector operations (like matrix product): https://eigen.tuxfamily.org/dox/group__TutorialArrayClass.html. In particular:
"Matrix expressions have an .array() method that 'converts' them into array expressions, so that coefficient-wise operations can be applied easily. Conversely, array expressions have a .matrix() method. As with all Eigen expression abstractions, this doesn't have any runtime cost (provided that you let your compiler optimize). Both .array() and .matrix() can be used as rvalues and as lvalues."

Reply:
That's exactly right---elementwise vs. matrix operations.  The conversions are efficient expression templates that share memory, so no copying and really no overhead because the wrappers get compiled out.